File: sample_spectra_calculations.m
% Spectral Analysis applied to georectified sample wave image stack

addpath _data/
addpath _functions/

load _data/angle_stack_x.mat
load _data/angle_stack_y.mat
load _data/freeboard.mat

slopefield_stack_x = tand(double(ax_int)*1e-3);
slopefield_stack_y = tand(double(ay_int)*1e-3);

fps = 15;
m_per_px_mean = 2*0.01074; % found from the Motion-Correction-Georeferencing Library

alt_mean = mean(freeboard);

mss_x = var(slopefield_stack_x,[],'all','omitnan');
mss_y = var(slopefield_stack_y,[],'all','omitnan');
mss = mss_x + mss_y;

slopefield_stack_x(isnan(slopefield_stack_x)) = 0;
slopefield_stack_y(isnan(slopefield_stack_y)) = 0;

Sx = double(slopefield_stack_x(:,:,1));
Sy = double(slopefield_stack_y(:,:,1));

% Select a region
[s1,s2] = size(Sx);
N_spect = 2^floor(log(min([s1 s2]))/log(2));
N_spect_2 = N_spect/2;


center_rows = floor(s1/2)-N_spect_2:floor(s1/2)+N_spect_2-1;
center_cols = floor(s2/2)-N_spect_2:floor(s2/2)+N_spect_2-1;

sx = slopefield_stack_x(center_rows,center_cols,:);
sy = slopefield_stack_y(center_rows,center_cols,:);

% COMPUTE SPECTRA FROM 3D STACK

w = circular_tukey(1+0*sx,0.2);
[dirspect,~] = compute_slope_spectrum(w.*sx,w.*sy,m_per_px_mean,fps,N_spect);
Skf = dirspect.Skf;
mean_heading =  0; % absolute heading is already considered in the rectification process, so the input is zero.
nframes = size(slopefield_stack_y,3);
df = fps/nframes; %fs/Nt
dk = 2*pi/(m_per_px_mean*length(center_rows)); % 2*pi/L_max
spectra_struc = compute_sub_spectra(Skf,dk,df,mean_heading,6);


% Figures
plot_sub_spectra(spectra_struc.f,spectra_struc.k,spectra_struc.nu,spectra_struc.theta,spectra_struc,1)

% Compare F_f and F_f obtained by F_k 
k = spectra_struc.k;
f = spectra_struc.f;
F_k = spectra_struc.F_k;
F_f = spectra_struc.F_f;
f_disp = sqrt(9.81*k)/(2*pi);
c = 2*pi*f_disp./k;
cg = c/2;
F_f_disp = F_k.*k./cg/(2*pi);
figure(101);loglog(f,F_f,'-',f_disp,F_f_disp,'-','linewidth',2)


% Plot k-f Spectrum both for Cross-Slope and Along-Slope
Skf_center = N_spect_2;

kvec = 0:dk:dk*2*Skf_center;
kvec = kvec - dk*Skf_center;
figure(998);imagesc(kvec,f,log10(squeeze(Skf(:,Skf_center,:)))');colorbar;view([0 -90]);xlim([-1 1]*100);hold on;plot(k,sqrt(9.81*k+0.072/1000*k.^3)/(2*pi),'-r',-k,sqrt(9.81*k+0.072/1000*k.^3)/(2*pi),'-r','linewidth',2);hold off
figure(999);imagesc(kvec,f,log10(squeeze(Skf(Skf_center,:,:)))');colorbar;view([0 -90]);xlim([-1 1]*100);hold on;plot(k,sqrt(9.81*k+0.072/1000*k.^3)/(2*pi),'-r',-k,sqrt(9.81*k+0.072/1000*k.^3)/(2*pi),'-r','linewidth',2);hold off

File: examples/sample_spectra_calculations.py
"""Spectral analysis applied to the georectified sample wave image stack.

Python port of sample_spectra_calculations.m: loads the bundled slope-angle
stacks, computes the wavenumber-frequency directional spectrum and all
sub-spectra, and reproduces the driver's figures.

    python examples/sample_spectra_calculations.py
"""

import sys
from pathlib import Path

import h5py
import numpy as np
import matplotlib.pyplot as plt

REPO = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(REPO))   # allow running without installation

from slopespectra import (compute_wave_spectra, figure_style,  # noqa: E402
                          plot_kf_slice, plot_omnidirectional,
                          plot_sub_spectra)
DATA = REPO / "_data"
FIGS = REPO / "_figures"

FPS = 15.0
M_PER_PX = 2 * 0.01074   # from the Motion-Correction-Georeferencing Library
GRAV = 9.81


def load_angle_stack(path, var):
    """(ny, nx, nt) slope stack from a MATLAB v7.3 milli-degree angle file."""
    with h5py.File(path) as h:
        ang = np.transpose(np.asarray(h[var]), (2, 1, 0))
    return np.tan(np.deg2rad(ang.astype(float) * 1e-3))


def main():
    FIGS.mkdir(exist_ok=True)
    sx = load_angle_stack(DATA / "angle_stack_x.mat", "ax_int")
    sy = load_angle_stack(DATA / "angle_stack_y.mat", "ay_int")

    # Heading is zero: absolute heading is already handled in rectification
    ds, dirspect = compute_wave_spectra(sx, sy, dx_m=M_PER_PX, fs_hz=FPS,
                                        heading_deg=0.0, n_freq_avg=3, temporal_alpha=0.1,
                                        n_nu_log=48, return_dirspect=True)
    print(f"mss full field = {ds.attrs['mss']:.4f}; analyzed crop = "
          f"{ds.attrs['mss_crop']:.4f}; spectral total = "
          f"{ds.attrs['total_slope_variance']:.4f}")
    print(f"k_cutoff = {ds.attrs['k_cutoff']:.1f} rad/m")

    figure_style()

    fig = plot_sub_spectra(ds)
    fig.savefig(FIGS / "sample_sub_spectra.png", dpi=200)

    # Compare F(f) with F(f) mapped from F(k) through linear dispersion
    ax = plot_omnidirectional(ds)
    ax.figure.savefig(FIGS / "sample_Ff_comparison.png", dpi=200)

    # k-f slices along both axes with the gravity-capillary dispersion curve
    fig, axs = plt.subplots(1, 2, figsize=(11, 4.5))
    plot_kf_slice(dirspect, axis="ky", klim=100, ax=axs[0])
    plot_kf_slice(dirspect, axis="kx", klim=100, ax=axs[1])
    fig.tight_layout()
    fig.savefig(FIGS / "sample_kf_slices.png", dpi=200)

    print(f"figures written to {FIGS}")


if __name__ == "__main__":
    main()
